### ISSUE TITLE
Consider adding `Content-Type: text/plain` to HTTP 500 response in `org.httpkit.server.HttpHandler#run`

### ISSUE THREAD

**Issue by sumbach** [http-kit/http-kit#559]
**Proposal:** Add `Content-Type: text/plain` header to the HTTP 500 response.

In the "exception handler of last resort" in `org.httpkit.server.HttpHandler#run`, any `Throwable` is automatically wrapped in an HTTP 500 response (great!) but the response doesn't include an explicit `Content-Type` header.

Security best practices appear to suggest including Content-Type in all (non-empty) HTTP responses. Missing Content-Type is much less concerning than errant Content-Type (anything with `Content-Type: text/html` is a potential XSS attack vector). I haven't yet been able to find any actual attacks/exploits caused by a missing Content-Type, though a cleverly constructed response body might trick some browser/user agent.

**Comment by sumbach** [comment]
I only noticed the "missing" Content-Type header because my ring middleware is only catching `Exception` and some non-`Exception` throwables were sneaking through to `org.httpkit.server.HttpHandler#run`.

Specifically, we had a failing `assert` (which throws `AssertionError`). It looks like `:pre` function preconditions also throw `AssertionError` as well.

Arguably, these should throw some appropriate `Exception` instead of `AssertionError` (which should be reserved for coding errors), but it's not implausible to encounter a non-`Exception` in the wild -- maybe `IOError`?

**Comment by ptaoussanis** [comment]
@sumbach Thanks for pinging about this Sam 👍

Have just added https://github.com/http-kit/http-kit/commit/36c8cdfc08a6bb65323102aa70d78a5af6950e9c - please take a moment to double check if you're happy with the approach there?

> Specifically, we had a failing assert (which throws AssertionError). It looks like :pre function preconditions also throw AssertionError as well.

Yes, this is unfortunately a bit of a common gotcha. Contrary to popular advice, I generally prefer catching `Throwable` for this reason. And it's also one of the reasons that I prefer throwing `ExceptionInfo`s for [Truss](https://github.com/taoensso/truss), etc.

**Comment by sumbach** [comment]
✨✨ @ptaoussanis ✨✨

This is **amazing!!** I just tested with a local http-kit build (commit e5b69e18b0216300ac51) and confirmed that my errant `AssertionError` now yields HTTP 500 with `Content-Type: text/plain; charset=utf-8`👍 💯 😀

The advice I've heard _against_ catching `Throwable` is that it's often pointless to catch and attempt to handle many JVM-initiated exceptions (like `OutOfMemoryError`) -- probably better to let the thread, and perhaps even the whole JVM process, crash on "unrecoverable" exceptions.

But when user code throws non-`Exception`, how can we distinguish between these and "unrecoverable" exceptions? I don't think we have a better choice than catching `Throwable`.

For my part, I will avoid `assert` (and `defn`+`:pre`?) for anything related to user input --
I'll take your suggestion to throw some appropriate unchecked `RuntimeException` (like `ExceptionInfo`).

**Comment by sumbach** [comment]
One comment on the approach: You could allocate and initialize the `HeaderMap` in an instance-level (or class static) constant.

I was thinking about the possibility of catching `OutOfMemoryError`, in which case you'd want to avoid allocations along this code path. But it looks like `HttpEncode` has to allocate `ByteBuffer`s (among other things), so there's probably no benefit to avoiding this `HeaderMap` allocation "just in case" we get an OOM.

That said, a constant `HeaderMap` (or dedicated factory function / constructor) could make this code _marginally_ more readable:

```clj
cb.run(HttpEncode(503, ContentTypeTextPlain, "Server unavailable, please try again", this.serverHeader))
```

**Comment by ptaoussanis** [comment]
> One comment on the approach: You could allocate and initialize the HeaderMap in an instance-level (or class static) constant.

Done, thanks Sam 👍 I've just released [v2.8.0-RC1](https://github.com/http-kit/http-kit/releases/tag/v2.8.0-RC1) which includes this work.

Cheers!

**Comment by sumbach** [comment]
Thank _you_, @ptaoussanis 😀 👍
